libyul/YulStack.cpp

```diff
@@ -230,7 +237,15 @@ YulStack::assembleEVMWithDeployed(optional<string_view> _deployName) const
 
 	evmasm::Assembly assembly(m_evmVersion, true, {});
 	EthAssemblyAdapter adapter(assembly);
-	compileEVM(adapter, m_optimiserSettings.optimizeStackAllocation);
+
+	// NOTE: We always need stack optimization when Yul optimizer is disabled (unless code contains
+	// msize). It being disabled just means that we don't use the full step sequence. We still run
+	// it with the minimal steps required to avoid "stack too deep".
+	bool optimize = m_optimiserSettings.optimizeStackAllocation || (
+		!m_optimiserSettings.runYulOptimiser &&
+		!yul::MSizeFinder::containsMSize(languageToDialect(m_language, m_evmVersion), *m_parserResult)
+	);
+	compileEVM(adapter, optimize);
```

Comment:
Note that here I'm preserving the ability to disable `yulDetails.stackAllocation` (but only when optimizer is enabled). Let me know if I should just permanently enable it and make the setting a no-op.